### PR TITLE
guice: mapBinder, aspects and singletons are implemented

### PR DESCRIPTION
**Pull request by qnislavnachev**
there is a small project that uses a guice api its about car interface
which got 3 implementations. In project are used a simple example of
multibindings, aspects and eager and lazy singletons.

### CODE REVIEW

**Comment by qnislavnachev** [comment]
PR is ready for review

**Comment by qnislavnachev** [comment]
little updates are made: no longer in provider cars are hard code

File: src/main/java/guicetasks/core/Person.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+
+import com.google.inject.Inject;
+import guicetasks.core.Car;
+
+public class Person {
+
+    @Inject
+    private Car car;
+
+    public Car getCar() {
```
Comment (mgenov) [inline src/main/java/guicetasks/core/Person.java]: The person is driving the car !

File: src/main/java/guicetasks/modules/LazyCarModule.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    @Override
+    protected void configure() {
+        MapBinder<Integer, Car> carMapBinder = getCarMapBinder();
+        OpelValidationInterceptor interceptor = new OpelValidationInterceptor();
+        bindInterceptor(any(), annotatedWith(CheckForOpel.class), interceptor);
+        bind(Car.class).toProvider(provider).in(Scopes.SINGLETON);
+    }
+
+    private MapBinder<Integer, Car> getCarMapBinder() {
+        MapBinder<Integer, Car> mapBinder = MapBinder.newMapBinder(binder(), Integer.class, Car.class);
```
Comment (mgenov) [inline src/main/java/guicetasks/modules/LazyCarModule.java]: MapBinder is not used correctly. Multiple modules should add there dependencies.

**Comment by qnislavnachev** [comment]
Used set as multi binder, used enums to remove the second module.

File: src/main/java/guicetasks/modules/MercedesProvider.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+import java.util.Set;
+
+public class MercedesProvider implements Provider<Car> {
+
+    public MercedesProvider() {
+        System.out.println("Instance was created");
+    }
+
+    @Inject
+    private Set<Car> multiBinder;
```
Comment (mgenov) [inline src/main/java/guicetasks/modules/MercedesProvider.java]: You have to inject Set of multibinder in a single place.

File: src/main/java/guicetasks/modules/CarModule.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        this.singleton = singleton;
+    }
+
+    @Override
+    protected void configure() {
+        Multibinder<Car> carMultiBinder = getCarMultiBinder();
+        OpelValidationInterceptor interceptor = new OpelValidationInterceptor();
+        bindInterceptor(any(), annotatedWith(CheckForOpel.class), interceptor);
+
+        if (Singletons.EAGER.equals(singleton)) {
```
Comment (mgenov) [inline src/main/java/guicetasks/modules/CarModule.java]: You can just show it with 2 different objects with logging in the constructor.

**Comment by qnislavnachev** [comment]
All task is reconstructed and using different objects to implements guice. There is 2 interfaces Person and HotDrinks  which got no login in it because  implementations of HotDrinks are used to show the differences between lazy and eager singleton.

File: src/test/java/guicetasks/AspectsTest.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        System.setOut(new PrintStream(out));
+        Injector injector = Guice.createInjector(new PersonModule(Maggie.class, Coffee.class));
+        AddressBook addressBook = injector.getInstance(AddressBook.class);
+        Person person = new FakePerson(123, "name");
+
+        addressBook.register(person);
+
+        assertThat(out.toString(), containsString("Registration is successful !"));
+        System.setOut(System.out);
```
Comment (mgenov) [inline src/test/java/guicetasks/AspectsTest.java]: System.out is already PrintStream, so this method is not doing anything.

File: src/test/java/guicetasks/AspectsTest.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    public void registerPersonWithNotNullId() throws Exception {
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        System.setOut(new PrintStream(out));
+        Injector injector = Guice.createInjector(new PersonModule(Maggie.class, Coffee.class));
+        AddressBook addressBook = injector.getInstance(AddressBook.class);
+        Person person = new FakePerson(123, "name");
+
+        addressBook.register(person);
+
+        assertThat(out.toString(), containsString("Registration is successful !"));
```
Comment (mgenov) [inline src/test/java/guicetasks/AspectsTest.java]: was successful

File: src/main/java/guicetasks/modules/PersonModule.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        bind(Integer.class).annotatedWith(Names.named("Jennie")).toInstance(333);
+
+        bind(HotDrinks.class).to(hotDrinksClass).asEagerSingleton();
+        bind(Person.class).to(personClass).in(Scopes.SINGLETON);
+
+        bind(Map.class).toProvider(StatisticsProvider.class);
+        bind(List.class).toProvider(ListOfAllPersonsNamesProvider.class);
+    }
+
+    private Multibinder<Person> persons() {
```
Comment (mgenov) [inline src/main/java/guicetasks/modules/PersonModule.java]: Doesn't look like a correct usage. The same thing could be represented as:

```java
@Provides
public Set<Person> getPersons() {
   return Sets.newHashSet(new Maggie(), new John(), new Jennie()); 
}
```

**Comment by qnislavnachev** [comment]
no longer using List provider in module now its implemented with set and using annotation @Provides and now it returns generic type

**Comment by qnislavnachev** [comment]
map binder example is included to the task and made some test for it

File: src/main/java/guicetasks/aspects/Register.java [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Target(ElementType.METHOD)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface
+Register {
```
Comment (mgenov) [inline src/main/java/guicetasks/aspects/Register.java]: badly formatted annotation

File: src/main/java/guicetasks/aspects/RegisterInterceptor.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+        Set people = (Set) field.get(book);
+
+        if (people.isEmpty() || !isContainPersonWithNoId(people)) {
+            throw new NullPointerException("Person ID cannot be null");
+        }
+        System.out.println("Registration was successful !");
+        return invocation.proceed();
+    }
+
+    private boolean isContainPersonWithNoId(Set<Person> people) {
```
Comment (mgenov) [inline src/main/java/guicetasks/aspects/RegisterInterceptor.java]: containsPersonWithoutId

**Comment by qnislavnachev** [comment]
chenges are made renamed method now its more understandable and formated annotation

File: src/main/java/guicetasks/modules/PersonModule.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+        bind(Integer.class).annotatedWith(Names.named("John")).toInstance(222);
+        bind(Integer.class).annotatedWith(Names.named("Jennie")).toInstance(333);
+
+        bind(HotDrinks.class).to(hotDrinksClass).asEagerSingleton();
+        bind(Person.class).to(personClass).in(Scopes.SINGLETON);
+
+        bind(Map.class).toProvider(StatisticsProvider.class);
+    }
+
+    private Multibinder<Person> getInstanceOfAllPersons() {
```
Comment (mgenov) [inline src/main/java/guicetasks/modules/PersonModule.java]: Doesn't looks use-full to be used like that as it could be replaced with a single module.

**Comment by qnislavnachev** [comment]
multibinder now its binded to Class not toInstance if I understood the comment correctly

**Comment by qnislavnachev** [comment]
implemented second module with his implementation of mapBinder and tested if it register all persons in address book

**Pull request merged**